File: llama-ui/src/utils/loadRules.ts
/**
 * TypeScript implementation of shared load rules between UI and backend
 */

export interface SystemLimits {
  maxContextSize: number;
  minContextSize: number;
  maxCpuThreads: number;
  memorySafetyMargin: number;
}

export const DEFAULT_SYSTEM_LIMITS: SystemLimits = {
  maxContextSize: 4096,
  minContextSize: 512,
  maxCpuThreads: 16,
  memorySafetyMargin: 0.2
};

export interface MemoryBreakdown {
  baseModelMb: number;
  kvCacheMb: number;
  computeBufferMb: number;
  outputBufferMb: number;
  safetyBufferMb: number;
}

export interface MemoryRequirements {
  totalMb: number;
  breakdown: MemoryBreakdown;
}

export class MemoryCalculator {
  static calculateKvCacheSize(contextSize: number, nLayers: number): number {
    const bytesPerToken = 2; // float16
    const kvCacheBytes = 2 * contextSize * bytesPerToken * nLayers;
    return kvCacheBytes / (1024 * 1024); // Convert to MB
  }

  static calculateTotalMemoryRequired(
    modelSizeMb: number,
    contextSize: number,
    nLayers: number,
    safetyMargin: number = DEFAULT_SYSTEM_LIMITS.memorySafetyMargin
  ): MemoryRequirements {
    // Base model memory
    const baseMemory = modelSizeMb;
    
    // KV cache
    const kvCache = this.calculateKvCacheSize(contextSize, nLayers);
    
    // Compute buffers (empirical measurements)
    const computeBuffer = 164.01;
    const outputBuffer = 0.12;
    
    // Calculate total with safety margin
    const subtotal = baseMemory + kvCache + computeBuffer + outputBuffer;
    const safetyBuffer = subtotal * safetyMargin;
    const total = subtotal + safetyBuffer;
    
    return {
      totalMb: total,
      breakdown: {
        baseModelMb: baseMemory,
        kvCacheMb: kvCache,
        computeBufferMb: computeBuffer,
        outputBufferMb: outputBuffer,
        safetyBufferMb: safetyBuffer
      }
    };
  }
}

export class LoadRules {
  static getValidContextSizes(
    minSize: number = DEFAULT_SYSTEM_LIMITS.minContextSize,
    maxSize: number = 32768
  ): number[] {
    const sizes: number[] = [];
    let size = minSize;
    while (size <= maxSize) {
      sizes.push(size);
      size *= 2;
    }
    return sizes;
  }

  static getRecommendedThreadCount(): number {
    const cpuCount = navigator.hardwareConcurrency || 4;
    // Use power of 2 up to 16 threads
    return Math.min(16, Math.pow(2, Math.floor(Math.log2(cpuCount))));
  }

  static validateContextSize(
    requestedSize: number,
    modelMaxContext: number,
    systemMaxContext: number = DEFAULT_SYSTEM_LIMITS.maxContextSize
  ): [number, string | null] {
    const validSizes = this.getValidContextSizes();
    
    // Find nearest valid size
    const nearestSize = validSizes.reduce((prev, curr) => 
      Math.abs(curr - requestedSize) < Math.abs(prev - requestedSize) ? curr : prev
    );
    
    // Apply limits
    const finalSize = Math.min(nearestSize, modelMaxContext, systemMaxContext);
    
    // Generate warning if adjusted
    let warning = null;
    if (finalSize !== requestedSize) {
      warning = `Context size adjusted from ${requestedSize} to ${finalSize} to match valid size and limits`;
    }
    
    return [finalSize, warning];
  }

  static canLoadModel(
    modelSizeMb: number,
    contextSize: number,
    nLayers: number,
    availableMemoryMb: number,
    safetyMargin: number = DEFAULT_SYSTEM_LIMITS.memorySafetyMargin
  ): [boolean, string] {
    const requirements = MemoryCalculator.calculateTotalMemoryRequired(
      modelSizeMb,
      contextSize,
      nLayers,
      safetyMargin
    );
    
    if (requirements.totalMb > availableMemoryMb) {
      return [
        false,
        `Insufficient memory. Required: ${requirements.totalMb.toFixed(1)}MB, Available: ${availableMemoryMb.toFixed(1)}MB`
      ];
    }
    
    return [true, "Model can be loaded with specified parameters"];
  }
} 

File: llama_cpp_api_package/utils/load_rules.py
"""
Load rules and validation for model loading.
Uses centralized configuration from model_config.py.
"""

from typing import Dict, List, Optional, Tuple
import psutil
import math
from llama_cpp_api_package.config.model_config import (
    SystemLimits,
    MemoryMultipliers,
    ModelDefaults,
    BufferSizes,
    MemoryCalculator as ConfigMemoryCalculator
)

class MemoryCalculator:
    """Memory calculation and validation"""
    @staticmethod
    def get_system_memory() -> Dict[str, float]:
        """Get system memory information in MB"""
        memory = psutil.virtual_memory()
        return {
            "total": memory.total / (1024 * 1024),
            "available": memory.available / (1024 * 1024),
            "used": memory.used / (1024 * 1024)
        }

    @staticmethod
    def calculate_total_memory_required(
        model_size_mb: float,
        context_size: int,
        n_layers: int,
        safety_margin: float = SystemLimits.memory_safety_margin
    ) -> Dict[str, float]:
        """Calculate total memory requirements using configuration"""
        return ConfigMemoryCalculator.calculate_total_memory_mb(
            model_size_mb=model_size_mb,
            context_size=context_size,
            n_layers=n_layers,
            safety_margin=safety_margin
        )

class LoadRules:
    """Rules for model loading and parameter validation"""
    @staticmethod
    def get_valid_context_sizes(
        min_size: int = SystemLimits.min_context_size,
        max_size: int = 32768
    ) -> List[int]:
        """Get list of valid context sizes (powers of 2)"""
        sizes = []
        size = min_size
        while size <= max_size:
            sizes.append(size)
            size *= 2
        return sizes

    @staticmethod
    def get_recommended_thread_count() -> int:
        """Get recommended thread count based on system CPU"""
        cpu_count = psutil.cpu_count(logical=True) or ModelDefaults.default_threads
        # Use power of 2 up to system limit
        return min(
            SystemLimits.max_cpu_threads,
            2 ** math.floor(math.log2(cpu_count))
        )

    @staticmethod
    def validate_context_size(
        requested_size: int,
        model_max_context: int,
        system_max_context: int = SystemLimits.max_context_size
    ) -> Tuple[int, Optional[str]]:
        """Validate and adjust context size"""
        valid_sizes = LoadRules.get_valid_context_sizes()
        
        # Find nearest valid size
        nearest_size = min(valid_sizes, key=lambda x: abs(x - requested_size))
        
        # Apply limits
        final_size = min(nearest_size, model_max_context, system_max_context)
        
        # Generate warning if adjusted
        warning = None
        if final_size != requested_size:
            warning = f"Context size adjusted from {requested_size} to {final_size} to match valid size and limits"
        
        return final_size, warning

    @staticmethod
    def can_load_model(
        model_size_mb: float,
        context_size: int,
        n_layers: int,
        available_memory_mb: float,
        safety_margin: float = SystemLimits.memory_safety_margin
    ) -> Tuple[bool, str]:
        """Check if model can be loaded with given parameters"""
        requirements = MemoryCalculator.calculate_total_memory_required(
            model_size_mb=model_size_mb,
            context_size=context_size,
            n_layers=n_layers,
            safety_margin=safety_margin
        )
        
        if requirements["total_mb"] > available_memory_mb:
            return (False, (
                f"Insufficient memory. Required: {requirements['total_mb']:.1f}MB, "
                f"Available: {available_memory_mb:.1f}MB"
            ))
        
        return (True, "Model can be loaded with specified parameters") 